src/main/java/io/reactivex/internal/operators/flowable/FlowableBufferBoundary.java

```diff
@@ -98,214 +119,340 @@ public void onSubscribe(Subscription s) {
         @Override
         public void onNext(T t) {
             synchronized (this) {
-                for (U b : buffers) {
+                Map<Long, C> bufs = buffers;
+                if (bufs == null) {
+                    return;
+                }
+                for (C b : bufs.values()) {
                     b.add(t);
                 }
             }
         }
 
         @Override
         public void onError(Throwable t) {
-            cancel();
-            cancelled = true;
-            synchronized (this) {
-                buffers.clear();
+            if (errors.addThrowable(t)) {
+                subscribers.dispose();
+                synchronized (this) {
+                    buffers = null;
+                }
+                done = true;
+                drain();
+            } else {
+                RxJavaPlugins.onError(t);
             }
-            actual.onError(t);
         }
 
         @Override
         public void onComplete() {
-            if (windows.decrementAndGet() == 0) {
-                complete();
-            }
-        }
-
-        void complete() {
-            List<U> list;
+            subscribers.dispose();
             synchronized (this) {
-                list = new ArrayList<U>(buffers);
-                buffers.clear();
-            }
-
-            SimplePlainQueue<U> q = queue;
-            for (U u : list) {
-                q.offer(u);
+                Map<Long, C> bufs = buffers;
+                if (bufs == null) {
+                    return;
+                }
+                for (C b : bufs.values()) {
+                    queue.offer(b);
+                }
+                bufs = null;
```

Comment:
shouldn't this be `buffers = null`?

Reply:
Yes, fixed. Also combined the error methods and changed to LinkedHashMap so that the final open buffers in `onComplete` remain in their order they were opened.